### ISSUE TITLE
BrokenPipeError leading to a crash when using filtering/counting shell utilities

### ISSUE THREAD

**Issue by fritschy** [borgbackup/borg#1116]
When using e.g. tail or head to limit the number of lines a BrokenPipeError is thrown and leads to an abnormal exit of borg:

```
% sudo borg list BACKUP | head -n1
20160202-121810                      Tue, 2016-02-02 13:03:51
Local Exception.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 1498, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 1452, in run
    return args.func(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 80, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 544, in do_list
    print(format_archive(archive_info))
BrokenPipeError: [Errno 32] Broken pipe

Platform: Linux occam 4.6.0-bait.0-wbt #1 SMP PREEMPT Tue May 17 09:32:06 CEST 2016 x86_64 
Linux: debian stretch/sid 
Borg: 1.0.2  Python: CPython 3.5.1+
PID: 21174  CWD: /home/user
sys.argv: ['/usr/bin/borg', 'list', 'BACKUP']
SSH_ORIGINAL_COMMAND: None

Exception ignored in: <_io.TextIOWrapper name='<stdout>' encoding='UTF-8'>
BrokenPipeError: [Errno 32] Broken pipe
```

***

:moneybag: [there is a bounty for this](https://www.bountysource.com/issues/34723987-brokenpipeerror-leading-to-a-crash-when-using-filtering-counting-shell-utilities)

**Comment by ThomasWaldmann** [comment]
We already had a similar case once and the issue that came up while fixing it was how to differentiate harmless BrokenPipeErrors (like head / tail / more / less) from not harmless ones.

**Comment by ThomasWaldmann** [comment]
Solving this might need to wrap screen output closely by an exception handler:

```
import sys

for i in range(100000):
    try:
        print(i)
    except BrokenPipeError:
        #might need to close stderr to avoid a warning emitted by python:
        #sys.stderr.close()
        sys.exit(0)
```

**Comment by enkore** [comment]
> Solving this might need to wrap screen output closely by an exception handler:

If so, then it should be done after console output was factored out from application logic.

**Comment by andela-kerinoso** [comment]
I would like to pick up this bug. I intend to combine suggestions from both @ThomasWaldmann and @enkore.

**Comment by ThomasWaldmann** [comment]
@andela-kerinoso ok, thanks! just keep us update via a pull request.

**Comment by ThomasWaldmann** [comment]
Hmm, PR #1151 is against master, this ticket is for 1.0-maint. Seems we need to decide.

**Comment by ThomasWaldmann** [comment]
See also the discussion in #1151.

**Comment by PlasmaPower** [comment]
Crazy idea, but would you be fine with patching stdout to handle a BrokenPipeError by closing the true stdout and ignoring future output?

**Comment by PlasmaPower** [comment]
BTW, I just double checked and that's possible. Example:

```python
import sys
import io

dev_null = open('/dev/null', 'w')
sys.stdout = io.TextIOWrapper(dev_null)

print("Hello no one!") # won't print anything
```

**Comment by PlasmaPower** [comment]
Oh, looks like we're already wrapping it. I'll see if I can't just switch around some stuff.

**Comment by ThomasWaldmann** [comment]
@PlasmaPower don't forget to start the bounty / claim it when this is closed.

**Comment by ThomasWaldmann** [comment]
Fixed by #1379 in 1.0-maint, soon to be also merged into master.
